Fe 102 add search and filters

GeorgDV opened the pull request:
Closes #102 

---
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

---

This adds a component that could be later used for searching and filtering products, since we dont yet have back-end data, it only console logs out the searchTerm string and filters array (currently as category ids array).

Demo with search string value and filter ids array console logged out:

https://user-images.githubusercontent.com/48998862/119643118-0f510480-be24-11eb-9697-f90eb5deaf5a.mp4

In mobile view:

![mobile](https://user-images.githubusercontent.com/48998862/119643192-22fc6b00-be24-11eb-80b6-7ec7aeb55ad1.JPG)

GeorgDV commented:
@Orzelius @tomimarkus991  Ready for review. See what you think about this.

tomimarkus991 commented:
That is really cool. Good job.

There are some things I would change and improve.
I think this looks better. It's a chakraUI component InputRightAddon and inside is SearchIcon.
![image](https://user-images.githubusercontent.com/48998818/119675649-e80b2f00-be45-11eb-8903-9735fd677f97.png)
![image](https://user-images.githubusercontent.com/48998818/119675903-17ba3700-be46-11eb-9f60-5a0d539379ab.png)

I noticed that if I try to press Enter it doesn't start searching. So user has to click The button everytime.
Also there should be atleast 2 char inside search input before the user can search. (this would help our backend 🚁)

Orzelius reviewed (commented):
I don't see this really working with how we have items/products set-up in the back-end right now. Most of this will have to be rewritten when switching to real data, so I think ti would make more sense to do it right now.

Currently in the back-end items and categories are working fine, so they can totally be used here as well.

GeorgDV commented:
Updated for now.

@tomimarkus991  I updated the styling a bit. Also now enter works for searching. Didn't add length checks yet, i think that could be added when the filters and search actually query items. Looks like this now:

![updated](https://user-images.githubusercontent.com/48998862/119978010-c5058a00-bfc1-11eb-8fb5-1f1c61273ee9.JPG)

@Orzelius I created a query to get categories and used it in `FiltersDropdown`. I don't see how i could query items yet, since `item` has no direct relation to `category` but `product` has, also pictures for fake data items should need sorting as well. I think that can be left to be fixed/added later though? 🤔 

What ya'll think?

Orzelius commented on fe/src/utils/category.ts:
```diff
@@ -0,0 +1,5 @@
+// @todo - Update this or remove after back-end implementation.
```
I think this can be removed now?

GeorgDV replied:
@Orzelius This seems kinda frustrating atm.  We could either use this custom type, or use no type at all [here](https://github.com/TPT-Loane/TPT-Loane/pull/108/files#diff-19e306ac23550d961cc0a799e8e282e7661c4ae12fd92638652a6419a72f13fdR17) (both technically work). 

The type generated by apollo has some specific property types set and couldn't be used right away. 🤔 

![type](https://user-images.githubusercontent.com/48998862/119984855-92ac5a80-bfca-11eb-8401-8937e0434b2a.JPG)
![typeerror](https://user-images.githubusercontent.com/48998862/119984856-9344f100-bfca-11eb-9a74-e34da7549469.JPG)

GeorgDV replied:
Currently i just removed it, see whatca think.

Orzelius replied:
We should only used gen-types. You can just return error or some kind of loading indicator if the data is undefined.

Orzelius replied:
```tsx
  const categories = categoriesQuery.data && categoriesQuery.data.categories;
  if (!categories) return <LoaderOrSmthng />
  categories.map(c => c); // <- this works becuase categories is no longer `| undefined`
```

GeorgDV replied:
Currently im just returning null if no categories are found, since we only display categories in dropdown atm. 

But for future reference i think some universal loading indicator component could be added to whereever we might need them. Actually a [spinner](https://chakra-ui.com/docs/feedback/spinner) is already provided by chakra. 🤔

Orzelius commented on fe/src/pages/Home/HomeContext.ts:
```diff
@@ -6,13 +6,21 @@ interface HomeContextType {
   setPage: React.Dispatch<React.SetStateAction<number>>;
+  searchTerm: string;
+  setSearchTerm: React.Dispatch<React.SetStateAction<string>>;
+  filters: Array<number>;
```
`number[]` is a shorthand for `Array<Number>` and generally is more often used

Orzelius commented on fe/src/components/SearchFilters/FiltersDropdown.tsx:
```diff
@@ -5,32 +5,16 @@ import {
 const FiltersDropdown: React.FC<Props> = ({ activeFilters }: Props): JSX.Element => {
   const { setFilters } = React.useContext(HomeContext);
+  const categoriesQuery = useGetCategoriesQuery();
+  const categories: Array<Category> = categoriesQuery.data ? categoriesQuery.data.categories : [];
```
If you don't feel like handling errors and load state in this pr, then at least log the error to the console if it exists.
